test/e2e/retry.go

```diff
@@ -33,17 +33,20 @@ const (
 	retryBackoffSteps           = 3
 )
 
-// retryWithTimeout retries the function until it returns true,
-// or a timeout is reached.
+// retryWithTimeout retries a function that returns an error until a timeout is reached
 func retryWithTimeout(interval, timeout time.Duration, fn func() error) error {
-	retryFn := func(fn func() error) func() (bool, error) {
-		return func() (bool, error) {
-			err := fn()
-			if err == nil {
-				return true, nil
-			}
-			return false, err
+	var pollError error
+	err := wait.PollImmediate(collectLogInterval, collectLogTimeout, func() (bool, error) {
+		pollError = nil
+		err := fn()
+		if err != nil {
+			pollError = err
+			return false, nil
 		}
+		return true, nil
+	})
+	if pollError != nil {
+		return pollError
```

Comment:
I'm still not sure why we need to check the pollError outside the poll function.

I found this example which might be closer to what you're trying to do: https://github.com/kubernetes/kubernetes/blob/4ab90ccebb0e22396b5e10cfd58bb34641bc766f/cmd/kubeadm/app/phases/upgrade/health.go#L163

```
if err != nil {
    return errors.Wrapf(pollError, "some text about failing after retries")
}
```

Reply:
The idea here is to bubble up the last error received if there is one. So the order of precedence:

1) return the most recent specific error from the function invocation, if it exists
2) if no error was ever returned from any function invocation, then return the error from the retry func (the error will either be nil or "timed out")

Does that make sense? The thinking here is that the most likely scenario if an error from the function invocation exists (inside PollImmediate) is a recurring, common error, so it aids to debugging to return that specific error so we can triage.

Reply:
I think I understand better now, thanks. Essentially, we don't want to return pollErr if `fn()` never returned and the retry just failed because of a timeout on the first attempt.

Reply:
Correct, in that case the "timed out" error *is* the fundamental error that we want to log/persist.